witnesslint.py

```diff
@@ -0,0 +1,1128 @@
+                    )
+            if self.programhash is None:
+                self.programhash = data.text
+            else:
+                self.log_with_position(
+                    LOGLEVELS["warning"],
+                    data.sourceline,
+                    "Found multiple definitions of programhash",
+                )
+        elif key == "architecture":
+            if self.architecture is None:
+                if data.text in ["32bit", "64bit"]:
+                    self.architecture = data.text
+                else:
+                    self.log_with_position(
+                        LOGLEVELS["warning"],
+                        data.sourceline,
+                        "Invalid architecture identifier",
+                    )
+            else:
+                self.log_with_position(
+                    LOGLEVELS["warning"],
+                    data.sourceline,
+                    "Found multiple definitions of architecture",
+                )
+        elif key == "creationtime":
+            if self.creationtime is None:
+                if not re.match(CREATIONTIME_PATTERN, data.text.strip()):
+                    self.log_with_position(
+                        LOGLEVELS["warning"],
+                        data.sourceline,
+                        "Invalid format for creationtime",
+                    )
+                self.creationtime = data.text
+            else:
+                self.log_with_position(
+                    LOGLEVELS["warning"],
+                    data.sourceline,
+                    "Found multiple definitions of creationtime",
+                )
+        elif key in self.defined_keys and self.defined_keys[key] == "graph":
+            # Other, tool-specific keys are allowed as long as they have been defined
+            pass
+        else:
+            self.log_with_position(
+                LOGLEVELS["warning"],
+                data.sourceline,
+                "Unknown key for graph data element: %s",
+                key,
+            )
+
+    def handle_key(self, key):
+        """
+        Checks a key definition for validity.
+
+        Should the key definition contain the mandatory 'id' and 'for' attributes the defined key
+        may be used in the appropriate data elements of any following graph definitions, even if
+        the key definition is faulty for other reasons.
+
+        Appropriate are all data elements that are direct children of an element of type
+        key_domain, which is the value of the 'for' attribute.
+
+        Key definitions in a witness may have a child element of type 'default' specifying
+        the default value for this key, but are currently expected to have no other children.
+        """
+        if "id" in key.attrib:
+            key_id = key.attrib["id"]
+        else:
+            self.log_with_position(
+                LOGLEVELS["warning"], key.sourceline, "Key is missing attribute 'id'"
+            )
+            key_id = None
```

Comment:
Applied this for every occurrence I found. Note that `None` is the default for the second parameter, so it can even be abbreviated to `key_id = key.attrib.get("id")`.